sdk/identity/Azure.Identity/src/CredentialPipeline.cs

```diff
@@ -57,5 +72,126 @@ public override bool IsRetriableResponse(HttpMessage message)
                 return base.IsRetriableResponse(message) || message.Response.Status == 404;
             }
         }
+
+        private class ScopeHandler : IScopeHandler
+        {
+            private readonly ClientDiagnostics _diagnostics;
+
+            public ScopeHandler(ClientDiagnostics diagnostics)
+            {
+                _diagnostics = diagnostics;
+            }
+
+            public DiagnosticScope CreateScope(string name) => _diagnostics.CreateScope(name);
+            public void Start(string name, in DiagnosticScope scope) => scope.Start();
+            public void Dispose(string name, in DiagnosticScope scope) => scope.Dispose();
+            public void Fail(string name, in DiagnosticScope scope, Exception exception) => scope.Failed(exception);
+        }
+
+        private class ScopeGroupHandler : IScopeHandler
```

Comment:
This type is solving the issue with #10659. Instead of creating inner scopes immediately, it stores required data and creates scopes based on the cumulative result. If authentication has succeeded, only one child scope is created.